src/object_store/src/object/opendal_engine/opendal_object_store.rs

```diff
@@ -0,0 +1,386 @@
+}
+#[derive(Clone)]
+pub enum EngineType {
+    Memory,
+    Hdfs,
+}
+
+impl OpendalObjectStore {
+    /// create opendal memory engine, used for unit tests.
+    pub fn new_memory_engine() -> Self {
+        // Create fs backend builder.
+        let mut builder = memory::Builder::default();
+
+        let op: Operator = Operator::new(builder.build().unwrap());
+        Self {
+            op,
+            engine_type: EngineType::Memory,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl ObjectStore for OpendalObjectStore {
+    fn get_object_prefix(&self, _obj_id: u64) -> String {
+        String::default()
+    }
+
+    async fn upload(&self, path: &str, obj: Bytes) -> ObjectResult<()> {
+        if obj.is_empty() {
+            Err(ObjectError::internal("upload empty object"))
+        } else {
+            self.op.object(path).write(obj).await?;
+            Ok(())
+        }
+    }
+
+    fn streaming_upload(&self, path: &str) -> ObjectResult<BoxedStreamingUploader> {
+        Ok(Box::new(HdfsStreamingUploader::new(
+            self.op.clone(),
+            path.to_string(),
+        )))
+    }
+
+    async fn read(&self, path: &str, block: Option<BlockLocation>) -> ObjectResult<Bytes> {
+        match block {
+            Some(block) => {
+                let range = block.offset as u64..(block.offset + block.size) as u64;
+                let res = Bytes::from(self.op.object(path).range_read(range).await?);
+
+                if block.size != res.len() {
+                    Err(ObjectError::internal("bad block offset and size"))
+                } else {
+                    Ok(res)
+                }
+            }
+            None => Ok(Bytes::from(self.op.object(path).read().await?)),
+        }
+    }
+
+    async fn readv(&self, path: &str, block_locs: &[BlockLocation]) -> ObjectResult<Vec<Bytes>> {
+        let futures = block_locs
+            .iter()
+            .map(|block_loc| self.read(path, Some(*block_loc)))
+            .collect_vec();
+        try_join_all(futures).await
+    }
+
+    /// Returns a stream reading the object specified in `path`. If given, the stream starts at the
+    /// byte with index `start_pos` (0-based). As far as possible, the stream only loads the amount
+    /// of data into memory that is read from the stream.
+    async fn streaming_read(
+        &self,
+        path: &str,
+        start_pos: Option<usize>,
+    ) -> ObjectResult<Box<dyn AsyncRead + Unpin + Send + Sync>> {
+        fail_point!("opendal_streaming_read_err", |_| Err(
+            ObjectError::internal("opendal streaming read error")
+        ));
+
+        let mut reader: ObjectReader = self.op.object(path).reader().await?;
+        let bytes = match start_pos {
+            Some(start_position) => {
+                let mut buf = Vec::new();
+
+                reader.seek(SeekFrom::Start(start_position as u64)).await?;
+                reader.read_to_end(&mut buf).await?;
+                Bytes::from(buf)
```

Comment:
This is not a good way to use the async reader since it basically reads data from the async reader synchronously. Given that `streaming_read` returns `AsyncRead` already, we can just return the reader directly without `seek`, `read_to_end` and `Bytes::from`.